File: python/src/insertion_sort.py
from typing import List

Numbers = List[int]

def insertion_sort(numbers: Numbers):
	sorted_numbers: Numbers = []

	for index, current_number in enumerate(numbers):
		is_first_iteration = index == 0

		if is_first_iteration:
			sorted_numbers.append(current_number)
			continue

		if len(sorted_numbers) == 1:
			if sorted_numbers[0] < current_number:
				sorted_numbers.append(current_number)
			else:
				sorted_numbers.insert(0, current_number)
		else:
			for sub_index, current_sorted_number in enumerate(sorted_numbers):
				if current_number <= current_sorted_number:
					sorted_numbers.insert(sub_index, current_number)
					break

	return sorted_numbers

if __name__ == "__main__":
	print(
		insertion_sort([78,36,46,43,31,0,3])
	)

File: typescript/src/insertion-sort.ts
function insertionSort(arr: Array<number>) {
	const sortedArr: typeof arr = []

	for (let i in arr) {
		const index = Number(i)
		const isFirstIteration = index === 0
		const currentNumber = arr[index]

		if (isFirstIteration) {
			sortedArr.push(currentNumber)
			continue
		}

		if (sortedArr.length === 1) {
			if (sortedArr[0] < currentNumber) {
				sortedArr.push(currentNumber)
			} else {
				sortedArr.unshift(currentNumber)
			}
		} else {
			for (let j in sortedArr) {
				const subIndex = Number(j)

				if (currentNumber <= sortedArr[subIndex]) {
					let subArr = sortedArr.splice(subIndex)
					sortedArr.push(currentNumber, ...subArr)
					break
				}
			}
		}
	}

	return sortedArr
}

console.log(
	insertionSort([23, 1, 10, 5, 2])
)